include/libopencm3/swm050/gpio.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2019 Icenowy Zheng <icenowy@aosc.io>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef LIBOPENCM3_GPIO_H
+#define LIBOPENCM3_GPIO_H
+
+#include <libopencm3/cm3/common.h>
+#include <libopencm3/swm050/memorymap.h>
+
+/* GPIO number definitions (for convenience) */
+/** @defgroup gpio_pin_id GPIO Pin Identifiers
+@ingroup gpio_defines
+
+@{*/
+#define GPIO0				(1 << 0)
+#define GPIO1				(1 << 1)
+#define GPIO2				(1 << 2)
+#define GPIO3				(1 << 3)
+#define GPIO4				(1 << 4)
+#define GPIO5				(1 << 5)
+#define GPIO6				(1 << 6)
+#define GPIO7				(1 << 7)
+#define GPIO8				(1 << 8)
+#define GPIO9				(1 << 9)
+#define GPIO_ALL			0x3ff
+/**@}*/
+
+/* GPIO direction definitions */
+/** @defgroup gpio_dir GPIO Pin Direction
+@ingroup gpio_defines
+@{*/
+#define GPIO_INPUT			0x0
+#define GPIO_OUTPUT			0x1
+/**@}*/
+
+#define GPIO_DATA			MMIO32(GPIO_BASE + 0x0)
+#define GPIO_DIR			MMIO32(GPIO_BASE + 0x4)
+#define GPIO_EXT			MMIO32(GPIO_BASE + 0x4c)
+
+#define GPIO_INTEN			MMIO32(GPIO_BASE + 0x30)
+#define GPIO_INTMASK			MMIO32(GPIO_BASE + 0x34)
+#define GPIO_INTLEVEL			MMIO32(GPIO_BASE + 0x38)
+#define GPIO_INTPOLARITY		MMIO32(GPIO_BASE + 0x3c)
+#define GPIO_INTSTATUS			MMIO32(GPIO_BASE + 0x40)
+#define GPIO_INTRAWSTATUS		MMIO32(GPIO_BASE + 0x44)
+#define GPIO_INTEOI			MMIO32(GPIO_BASE + 0x48)
+
+#define SWD_SEL				MMIO32(SYSTEM_CON_BASE + 0x80)
```

Comment:
Should this define exist? it's got a prefix that doesn't match this file, and it's a duplicate of the next three defines.  (likewise, why do you need all four of these?)

Reply:
oh sorry, copy-pasta error, should be 0x30, 0x80, 0x90, 0xe0.

For the SWD_SEL logic, the two pins have 3 status possible: SWD, AF and GPIO. When SWD_SEL is 1, SWD is available; when SWD_SEL is 0 and GPIO_SEL (in the datasheet named PORTA_SEL, but this chip has only Port A) have the AF of the corresponding pin set, AF is available, otherwise GPIO is available.